### DISCUSSION TITLE
Adding a column using EF code first HELP Please

### DISCUSSION THREAD

**Discussion by sisaacks** [Azure/azure-mobile-apps#350] [Q&A]
Hello, 

I have an Xamarin shell application.  I followed the online tutorial for Syncing Data from the onboard Sqlite DB to an Azure SQL server in the cloud.  The data synced just as it should.  

https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/client/dotnet

I then added a column to my table on the Server Database.  I ran the EF migration and populated the column, it worked flawlessly 
I then went back to the client and added the property to my Model
Now when I try to sync as soon as I call await MyTable.PushItemsAsync();
I get an internal server error 500
Everything was working perfectly before I added the column
So I am not sure why its breaking now or how to figure it out?  
Any help would be much appreciated.  

This is the error I am getting:  {"type":"https://tools.ietf.org/html/rfc7231#section-6.6.1","title":"An error occurred while processing your request.","status":500,"traceId":"00-2324ac47ca84d68f2125236838aea85d-49f03b0d5eb0a4a0-00"}

I cleared the data cache for the app on the phone, ran the app again and now it works. 
I then added another property and ran into a diff error

"{\"type\":\"https://tools.ietf.org/html/rfc7231#section-6.5.1\",\"title\":\"Bad Request\",\"status\":400,\"trac…"

I am obviously doing something wrong.  My controller is just the basic Table Controller from the example...is there something I Need to do different on the controller?

**Comment by adrianhall** [comment]
First off, add some logging to your client application: https://github.com/Azure/azure-mobile-apps/wiki/How-to-use-the-.NET-Client-SDK#enable-request-logging - see the requests going in and out.

Secondly, add some logging to your service application: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0

The problem you are running into is likely a mismatch between what is in queue on the client (i.e. what the client is sending) and what the server says it needs for a valid model.  You've not provided the models on either side and I don't know how your SQL server is set up.  You need to understand a little bit about REST to do this so you can take the client out of the picture and send raw requests and see what the response is.  That then isolates the problem to one place or another.

**Reply by sisaacks** [reply]
@adrianhall The Azure Sql Database is just a normal database i set up through the tutorial.  I guess what confuses me is I followed the tutorial to a T its syncs perfectly until I add a column on the server and client.  I can't be the first person that this has happened too.  
It seems like something trivial...that its a step I am missing.  

I deleted both databases, started over, added another property and now I am getting a 400 error bad request.  Again it worked fine before I added the property.  The controller is the basic controller in the tutorial.  

So in the models below, when I first created the DB on the client and server, "Doors" was not there.  
I added Doors to the server, migrated, and populated the column.  
I then went back to the client, added Doors to the client model
I updated the Object on the client to have 2 doors as I had done on the server.
I Then tried to sync and I got the 400 bad request. 

**Client Model**
```
[DataTable("vehicles")]
    public class Vehicle : DatasyncClientData
    {
        //public Guid Id { get; set; }
        public string Make { get; set; }
        public string Year { get; set; }
        public string Color { get; set; }
        public string VehicleModel { get; set; }

        public VehicleType VehicleType { get; set; }

        public int Doors { get; set; }
    }
```

**Server Model**
```
public class Vehicle : EntityTableData
    {
        public string Id { get; set; }
        public string Make { get; set; }
        public string Year { get; set; }
        public string Color { get; set; }
        public string VehicleModel { get; set; }
        public VehicleType VehicleType { get; set; }
        public int Doors { get; set; }
    }

    public enum VehicleType
    {
        Car = 1,
        Truck = 2,
        SUV = 3
    }
```

**Reply by adrianhall** [reply]
You need to drop the table on the client and re-create it.  Each and every time.

The way I do this is that I have a database version on the client (embedded).  I have an ApiController on the service that returns the version of the database on the service.  If the database versions differ, I pop up a window telling the user to upgrade the app.  I then store the database version in settings on the client.  If the embedded version and the version in settings are different, I delete the offline database and do a full create a re-sync.  Generally, I have a page for this, so my app goes Splash => Database Setup => Authentication => App

**Comment by sisaacks** [comment]
Adrian

Thank you for the input.  It leaves me with a question, what if the mobile devices has data that has not been synced to the server yet.  Dropping the database would lose the data.  I ask because a large portion of the time our users will be working offline. 

Seth

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Reply by adrianhall** [reply]
This is where it gets really complex.  The best way is to keep the old version around (so you are running two versions of the backend - version 1 and version 2).  You update version 1 so it accepts the old model version, but writes the new model version, and returns the old model version when a pull happens.

When your client connects and determines it needs to update, you:

* Push to version 1 
* Delete the database and re-setup as version 2.
* Pull all the records from version 2.

I'm covering update and production cloud stuff a lot more in the upcoming book.

**Comment by sisaacks** [comment]
I kind of follow what you’re saying. Trying too digest it.  So are you saying to set up two offline stores? Anyway you could send me a snippet of what that looks like?  How does the client know which store too pull records from after the update? 

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Comment by sisaacks** [comment]
Adrian Also the failure happens on  await vehicleTable.PushItemsAsync(); so
pushing to VErsion 1 would cause the error.
You said you have a book coming out?
Any help you can provide would be appreciated....I am kind of on a time
crunch to get this working.
Have a good weekend.

Seth

**Comment by sisaacks** [comment]
I dont see a way to drop the table or the database using the store or client….nor a way to get the sqlite db version.  I must be googling the wrong thing, i would think others would have had to deal with this.  
Everything will be perfect as soon as I learn how to do what you’re saying.  

Im confused how version1 and version2 would be any different being they are both based on the same model table.  This is where i get lost.  

Maybe I am doing things out of order? 
1.  Update the server model
2.  Migrate with EF
3.  Update the SERVER db
4.  Update the client model with the new property

Now run the app and sync.  The first push breaks.  From my understanding because the versions are diff? 

Being Im manually adding properties to the model not sure how version 1 and version 2 would be any different.  

Sql-net-pcl its easy to drop tables, but I am unsure how to do it with this? I think the only way you can delete the db is too delete the file.  

Sorry about all the questions, i am spinning my wheels at finding info on this

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Reply by fileman** [reply]
have you tried to purge all items in client table before push from server?
`await _offlineTable.PurgeItemsAsync_offlineTable.CreateQuery(), new PurgeOptions { DiscardPendingOperations = true});`

if it's not a problem for you to delete the db file, it might be an option.
I have added this features to #324

**Reply by sisaacks** [reply]
@fileman 
so I am trying to do what you suggested above, however I do not seem to have the options.
var vehicleTable = datasyncClient.GetOfflineTable<Vehicle>();
vehicleTable.PurgeItemsAscyn does not have CreateQuery()

**Comment by sisaacks** [comment]
Adrian

Sorry to bug you again, I did find your book:
https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/
However it doesn't seem to cover the issue I am having.
Hopefully you can shed some light on this.

Regards
Seth

**Reply by fileman** [reply]
the link is for the older version on this library and is very different form the current implementation.

**Comment by sisaacks** [comment]
Ok where can I find the most current link?

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Reply by adrianhall** [reply]
The new book hasn't been written yet.

**Comment by sisaacks** [comment]
Well can anyone help with the issue Im having? Of
Explain what I am doing wrong?

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Comment by sisaacks** [comment]
The issue with purging, the client would lose any data that has not been updated too the server.  But i will try what you suggested to see if it works

SUA SPONTE
-- Its better to fail while daring greatly,  than to be with those cold and timid souls who neither know victory nor defeat.
-- Someone once told me not to bite off more than I could chew. I said I'd rather choke on greatness than nibble on mediocrity.

**Reply by fileman** [reply]
you don't lose client side changes and fetch server side, but step are different with / out SoftDelete

Without SoftDelete the steps should be:
1. Push items changed to server
2. Purge all items (offline table contains even deleted items on server side)
3. Pull new items in offline table

With SoftDelete the steps should be:
1. Push items changed to server
2. Pull updated version of items from server
3. Purge "soft" deleted items in offline table

**Reply by sisaacks** [reply]
@fileman 
The problem is, after I have changed the schema on the server and client, the first push (listed in step 1) I get the error message I have been posting about.  
I dont know how to update the schema without getting the mentioned error.....
as i stated I have an Xamarin Cross Platform Application using sqlite-net-pcl
Our clients work offline mostly
when I update the schema on the server using Entity Framework MIgration
then add the updated property to the client, sqlite-net-pcl creates the updated table
when I try to do the first step you listed....is when I get the error.  
I can't seem to get any help with this issue or a valid example of how to accomplish this

**Reply by adrianhall** [reply]
I'm not sure how we can be more clear.  You need to either adjust the SQLite database and definitions before you start OR you need to detect the change (outside the scope of the framework), drop the table, and re-sync.  There are methods on the SQLiteStore for doing random SQLite database calls (execute and query) that you can use to effect changes.

We've suggested multiple ways to deal with this.  It's up to you to actually code those changes.

**Reply by sisaacks** [reply]
@adrianhall how can you be more clear?  Adrian I am not a DBA.  I have looked through  your book and online and I do not see how to delete the database other than delete the file?  Is that what you mean?  Sometimes you say delete the database and sometimes you say the table....so I am confused as to which you mean?   

You said that you check the version numbers...how do I get the version numbers?  I do not see that in the documentation anywhere.  Unless you referring the the version number in the tables?  I checked those before I did the update and the sqlite and SQL version number are different.  

@adrianhall You mentioned keeping two databases on the client, to which I replied and asked for an example, because I am not sure how that would work or how to do that.   You said this is the ONLY way I would not lose data.  I am not sure how to do what you said though....can you PLEASE give me a little more insight?  I am not looking for anyone to write the code for me, I am looking to understand what I need to do...and right now I do not.  

What do you mean I need to adjust the sqlite database and definitions before I start?  What does that mean?  

If I NEW how to do it I would not still be asking.  I am missing something that is OBVIOUS to you but NOT to me.

**Reply by fileman** [reply]
@sisaacks are you using sqlite-net-pcl instead of library OfflineSQLiteStore?
Can you share your code in gist or private repo, otherwise it is too difficult to help you.

**Reply by fileman** [reply]
@sisaacks dont use #324 for this issue.
If I have understand @adrianhall's reply, you must have 2 "backend controller" using different entity schema. 
The client check version, if there is a newer, push all V1 items to the older controller and then delete db file. 
After new client db is initialized with V2 schema table pull all items.
This way to keep older client working without breaking changes.

**Reply by sisaacks** [reply]
@fileman I am using OfflineSQLiteStore.  
I am not sure how to publish to a private repo.
There is not much to my code, its a very simple example.  
Essentially when the app starts this  happens 

```
var dbPath = $"{FileSystem.AppDataDirectory}/NFO_DB.db3";
store = new OfflineSQLiteStore($"file://{dbPath}?mode=rwc");
store.DefineTable<Vehicle>();

options = new DatasyncClientOptions
{
   OfflineStore = store
};

datasyncClient = new DatasyncClient("https://autowebapi2022.azurewebsites.net", options);
await datasyncClient.InitializeOfflineStoreAsync();
```

Then I add a object on the client and sync to the server.
The sync works as it should

Then I use EF to add a Property to the model on the server
Then I go back to the client and add the property in my model
Then I try to sync again and when I call the first PUSH that is when it pukes 

```
var vehicleTable = datasyncClient.GetOfflineTable<Vehicle>();
await vehicleTable.PushItemsAsync();
await vehicleTable.PullItemsAsync();
```

**Reply by sisaacks** [reply]
@fileman again I am not sure how that would possibly work with different schema....because  the schema is based off of the model....and if I add a property to the model, then the schema will change for both.  There is something I am obviously missing in the explanation.  I am not looking for anyone to write the code for me, I am trying to understand what needs to be done.  I can't be the only person that has had to do this.  

I am also confused if you guys are talking about the version number in the Tables or the DB version number.  After the first sync, I can see the version for the Vehicle table...and the version on the client and server is already different.  

The version on the server shows: 0x00000000000007D1
The version on the client shows:  AAAAAAAAB9E=

**Reply by fileman** [reply]
@sisaacks I think adrian meant entity schema not to be confused with database schema (ex. dbo)
have you tried:
```
var vehicleTable = datasyncClient.GetOfflineTable<Vehicle>();
await vehicleTable.PushItemsAsync();
await vehicleTable.PrugeItemsAsync(vehicleTable.CreateQuery(), new PurgeOptions { DiscardPendingOperations = true });
await vehicleTable.PullItemsAsync();
```

**Reply by sisaacks** [reply]
@fileman 
I tried the following...but the vehicle I added was lost...but it did not error out.  The property I added is "Doors".  If I do not purge first, then when I call the first PUSH it will error out.  However calling purge first I loose any items recently added on the client that have not been synced 

```
var vehicleTable = datasyncClient.GetOfflineTable<Vehicle>();

Vehicle jeep = new Vehicle
{
   Id = Guid.NewGuid().ToString(),
   Color = "Green",
   VehicleModel = "Rubicon",
   Make = "JEEP",
   Year = "2019",
    VehicleType = VehicleType.SUV,
    Doors = 2
};

await vehicleTable.InsertItemAsync(jeep);


await vehicleTable.PurgeItemsAsync("", new PurgeOptions {DiscardPendingOperations = true });
await vehicleTable.PushItemsAsync();
await vehicleTable.PullItemsAsync();
```
However now if I run the app again and add the vehicle back...then sync...it works perfect.  So I know I am close....there is something I am missing though

**Reply by fileman** [reply]
wrong order!
with client/server same entity version, it should be:
1. Insert -> add new vehicle
2. Push -> send to server
3. Purge -> clear local copy
4. Pull -> get from server

to clear client db with older entity version and single backend with new version
1. Purge
2. Pull
3. Insert
4. Push

**Reply by sisaacks** [reply]
@fileman  performing the 'to clear client db with older entity version and single backend with new version'
wont i lose all unsynced data on the local?

**Reply by fileman** [reply]
yes, since backend cant handle older you have to clear all.

Instead you can use asp.net core versioning to have both controller for older and newer entitie's version, have you tried?

**Reply by sisaacks** [reply]
@fileman I don't know how to do the versioning you're speaking of.  I am pretty new to web.  I didnt realize this was going to be so problematic to add a column....I doubt it will happen much...but in the event that it does, I need to figure out how to deal with this.

**Reply by sisaacks** [reply]
@fileman can you tell me  how to delete a table using the framework?
I tried the following and it will not work
 ```
var dropTableStatment = "DROP TABLE vehicle";
await store.ExecuteQueryAsync(vehicleTable.TableName, dropTableStatment);
```
Cannot prepare statement for 'DROP TABLE vehicle'
  at Microsoft.Datasync.Client.SQLiteStore.Driver.SqliteConnection.PrepareStatement

**Reply by sisaacks** [reply]
@fileman everything I have tried fails.  Either I get an error or it will not update the new column. 
@fileman I follwed your instructions to a T and the column does not update on the server, I get the default value every time. There is no supporting documentation on your site for how to do this.  Unfortunately I have to move on and this was a complete failure. 

For future reference, for beginners, I have no idea what you mean by version.....there are multiple versions out there....table, column and database.
It would also be nice to know how to delete an offline table, that is not in the documentation
it would also be nice to know how to deal with adding/removing columns once its beeen synced...like a detailed example in the documentation.  
It might be second nature to you guys, but to us, those who are new to this wheel house, its unknown.

**Reply by sisaacks** [reply]
@adrianhall @fileman Can someone explain to me how to drop a offine table? I have tried below but I get an error stating it cannot prepare this statement.

var myOffLineTable= datasyncClient.GetOfflineTable<MyTablle>();
var sqlStatement = "DROP TABLE myOffLineTable";
await store.ExecuteQueryAsync(myOffLineTable.TableName, sqlStatement);
I also deleted the local db file.
Created a new store and client
Pulled
Inserted
Pushed

It did not error but the column did not update the value on the server, but it did on the client. Has anyone successfully done this in an Xamarin application?

**Reply by fileman** [reply]
I think you can't do a "DROP TABLE", but I'm not sure, you have to wait @adrianhall reply for this.
at now I have no time to create a client/server solutions to test your issue, can you create it and put in repository to share?

**Reply by sisaacks** [reply]
@fileman First off...THANK YOU.  Can I zip my project and sent it to you on here?  If I have to delete the entire DBA file and then repopulate a new sqlite DBA...that kind of stinks.....because any current local changes/new that have not been pushed would be lost, unless I manually repopulate data.  I would think that someone doing an Xamarin application would have done this.  But I have found nothing googling.  I will gladly zip this project and send it to you.

**Reply by sisaacks** [reply]
@fileman I got your invite, but wont you need access to the backend azure server as well?  I mean the issue happens with you update the schema on the server.  I will gladly send you the project. 

I finally got it to work.....well kind of.  Essentially this is what I have to do upon a schema change

Step 1:  Get all the tables and data from the current store such as 
var vehicleTable = datasyncClient.GetOfflineTable<Vehicle>();

Step 2:  Dispose the current Store

Step 3:  Delete the current database file

Step 4:  Create a new store and define new database tables

Step 5:  Insert the old data from the old tables into the new tables 

Step 6:  Push

Step 7:  Pull

This does work, but for any items already existing on the server I get an error because the new added column has a different value I am assuming.  I am guessing I need to do some kind of update for that row that is already existing, or maybe its because the schema is different for that row....I am not sure.

**Reply by fileman** [reply]
no, I dont need azure access, just see your code for overview and replicate your use case

**Reply by sisaacks** [reply]
@fileman  
I uploaded the files

@fileman 
I think the issue I mentioned above, about the rows already on the server causing the error. I think may be due to schema mismatch?  Not sure.  But it seems like quite a pain to have to delete everything and repopulate?  I truly appreciate your assistance in this endeavor

**Reply by adrianhall** [reply]
My suspicion at this point is that you have a new REQUIRED field, which is causing the error - the validation fails.  Capturing the HTTP transaction would help in diagnosis.

You can drop an entire table with store.Execute* methods - there is one for queries and one for non-queries.

**Reply by adrianhall** [reply]
Specifically, you can use 

``` csharp
        protected void ExecuteNonQueryInternal(string sqlStatement, Dictionary<string, object> parameters = null)
```

Maybe I should provide an external version of this.  Right now, you should subclass the OfflineSQLiteStore to be your own type, then you can create a new method that calls this one.

``` csharp
public class MyOfflineStore : OfflineSQLiteStore 
{
  public void ExecuteNonQuery(string sqlStatement, Dictionary<string, object> parameters = null)
    => ExecuteNonQueryInternal(sqlStatement, parameters);
}
```

Then you can use:

``` csharp
var store = new MyOfflineStore(connectionString);
store.ExecuteNonQuery("DROP TABLE myTable");
```

**Reply by sisaacks** [reply]
@adrianhall I am supplying a value in the required field why would it fail?  this is so frustrating

**Reply by sisaacks** [reply]
@fileman 

If you have any luck please let me know.  From my standpoint there does not seem to be enough documentation for the issue I am describing.  The issue where the user has changes on the client that have not been synced yet prior to the database being updated. 
From what I am seeing this is what needs to happen which seems odd being this is a sync framework

1.  Get the data from the current tables
2. Delete the database file
3. Create a new store and data sync client
4. Find out what data this on the client that has been changed and not synced.....this data will have to be manually updated on the server using ReplaceItemAsync.  Any data that is new data, on the client and not on the server will have to be added back to the client table via the InsertItemAsync method .  Then you can do a Push and Pull.  

I thought the above would work, however when I queried the Remote table, got the ONE item in the remote table, updated the item and then tried to replace the item, I got a "bad request".  Which seems rather confusing, being that the item was handed to me from the server but when I try to give it back, its a bad request.  Again, followed the instructions and online example.  

I then tried to update on the client and then push the record to the server.....same thing...for the record that was being updated on the server.....bad request.  So there seems to be an issue with the "update" on the server.  I thought maybe it was something to do with azure, so I did another regular api with CRUD and it worked fine.  

I know you guys are busy....but not sure whats going on...when I followed the documentation that is available 

Does the controller need to be any different than the example for doing an update or replace?  I was told there would be new documentation and examples last week but I have yet to see anything.

**Comment by ideas2appsaustralia** [comment]
@sisaacks 

I have done an add-migration to put a new column in to a table in the backend server sql database.  SyncOfflineCacheAsync and GetTableAsync does not return any data from  the new column of teh table. only data from the columns that were in the table when I created offline tables and dbase initially are returned . 

I tried @fileman suggestion above ie

`                foreach (var offlineTable in offlineTables)
                {
					await offlineTable.PushItemsAsync();
					await offlineTable.PurgeItemsAsync("", new PurgeOptions { DiscardPendingOperations = true });
					await offlineTable.PullItemsAsync("", options);
                }
`

with no success.

In my logs I can see the HTTP where items are returned and it does not include the new column at all.

Did you find a solution different to above?

**Comment by sisaacks** [comment]
Scott

I never got it to work.  I ended up creating my own sync.  I needed
something very simple.  The onboard DB, Sqlite, is the primary DB in my app
being that our users will be offline mostly.
They just need the ability to login and sync data when there is service.  I
spent hours trying to get the above to work.  I found a simply sync
algorithm implemented by a guy on the Xamarin team and it works great.

**Comment by adrianhall** [comment]
Make sure you check to ensure the SERVICE is returning the columns - use Postman to do a GET /tables/{your table} and make sure the data is returned.  No amount of wishing will make the data exist if the server is not returning it.

This now means you can tell whether it's the service or the client that is the problem.

If it's the service, then ensure (a) you've done a migration, (b) the migration is applied (i.e. look at the database table definition in the database itself),  (c) you've updated the model for the table, and finally (d) you've updated the existing records to add any missing data.  If all that has been done, the data should be flowing from the service.

If it's the client, then wiping the database, changing the model and re-creating the database will ensure the data is updated from the service.  Technically, the database will be updated by any non-breaking (i.e. additive) change to the model.  However, if that doesn't happen for some reason, you need to do it manually, which means wiping the database and re-synchronizing, unfortunately.

My bet is that the service changes didn't actually happen.

**Comment by ideas2appsaustralia** [comment]
So your a betting man are you? And do I read between the lines that you suspect its my coding ability ?

Well you'd be right on both accounts!

It was the backend service changes not actually happening. The azure sql database was fine, it had the new column, all the migrations worked, the data was there, but some idiot had not re-published the API to Azure with the new model!!!

**Reply by adrianhall** [reply]
I've just seen enough problems to figure out the most likely place to look.  There is no replacement for experience.

**Comment by sisaacks** [comment]
I am not sure what I was doing wrong, but maybe its because of how I am
creating the SQLite database.   The onboard will be our primary database,
the server is just a backup.
Most of the time our users will be offline.  Whenever I updated a column in
a table, I would always get an error message stating that a tracking table
didnt exist.
I finally gave up and created my one simple sync and it works great.
